presto-main/src/main/java/com/facebook/presto/operator/scalar/QuantileDigestFunctions.java

```diff
@@ -112,4 +161,66 @@ public static long verifyWeight(long weight)
         checkCondition(weight > 0, INVALID_FUNCTION_ARGUMENT, "Percentile weight must be > 0, was %s", weight);
         return weight;
     }
+
+    public static void validateQuantileRange(double lowerQuantile, double upperQuantile)
+    {
+        checkCondition(0 <= lowerQuantile && lowerQuantile <= 1, INVALID_FUNCTION_ARGUMENT, "lowerQuantile must be between 0 and 1");
+        checkCondition(0 <= upperQuantile && upperQuantile <= 1, INVALID_FUNCTION_ARGUMENT, "upperQuantile must be between 0 and 1");
+        checkCondition(lowerQuantile < upperQuantile, INVALID_FUNCTION_ARGUMENT, "lowerQuantile must be strictly less than upperQuantile");
+    }
+
+    /**
+     * A histogram made from two quantiles returns two Buckets:
+     * 1) values below the lower bound (exclusive)
+     * 2) values between the lower bound (inclusive) and the upper bound (exclusive)
+     * <p>
+     * The truncated mean comes from the second Bucket
+     * If the lower and upper bounds are the same, the Bucket is empty so we return null
+     */
+    public static Double digestToTruncatedMean(QuantileDigest digest, List<Long> bounds, QuantileDigest.MiddleFunction middleFunction)
+    {
+        if (bounds.get(0).equals(bounds.get(1))) {
```

Comment:
Because we validate the range above, can we change this into a preconditions check where we assert that the first element is lower than the second?

Reply:
Added as a code comment:
> even if the quantiles are different, the corresponding values that constitute the histogram bounds may be the same, e.g. in the set `(1,5,9)`, quantiles 0.49 and 0.5 both resolve to the value 5.

Reply:
What if we have a qdigest like so

`(1,2,2,2,2,2,2,2,2,2,3)`

And our quantiles are 0.3 and 0.6.

Reply:
Good point. I'll change it so it returns `2` in this case.

Reply:
Actually I'm not sure it's possible to accurately detect this scenario from the qdigest.
If the bounds are the same, we would essentially need to determine how many values lie in the quantile range, but that seems impossible if some values are weighted.

The problem is, truncated mean is ideally computed by truncating based on quantile, but with the qdigest->historgram approximation, we use values at quantiles, rather than quantiles, to do the truncation, so there will always be edge cases when we encounter repeated values at the quantile bounds. In fact this is true whether the lower and upper bound values are the same or not (using your qdigest example, if the quantiles were 0.0 and 0.6, the trunc mean would be 1.0 because none of the 2s would be in the histogram).

I think the best approach is to just be clear that this limitation exists. What do you think?

Reply:
It would be much more ideal to get it work correctly for this scenario.  I think we could do that in a couple of ways, both of which involve changes in the underlying library we use:

1) Add a public method to traverse the qdigest tree
2) Implement truncated means on the qdigest itself and use it here

I created https://github.com/airlift/airlift/issues/710 to gauge interest in the former approach.

Reply:
Sure, that makes sense. Thanks for making that issue. I'll have a go at implementing the truncated mean directly on the q-digest unless you are already working on that?

Reply:
I think it makes sense to implement option 1 in Presto, provided the public method is exposed.  I created a gist of how it might look, but it's not rigorously tested: https://gist.github.com/tdcmeehan/b09866f4eab430c346a97a305fdb2252

Reply:
Awesome, thanks for that. So I guess we wait until Airlift allows us to expose that traversal.